app/code/Magento/Quote/Model/Cart/AddProductsToCart.php

```diff
@@ -0,0 +1,216 @@
+     */
+    private $maskedQuoteIdToQuoteId;
+
+    /**
+     * @var BuyRequestBuilder
+     */
+    private $requestBuilder;
+
+    /**
+     * @param ProductRepositoryInterface $productRepository
+     * @param CartRepositoryInterface $cartRepository
+     * @param MaskedQuoteIdToQuoteIdInterface $maskedQuoteIdToQuoteId
+     * @param BuyRequestBuilder $requestBuilder
+     */
+    public function __construct(
+        ProductRepositoryInterface $productRepository,
+        CartRepositoryInterface $cartRepository,
+        MaskedQuoteIdToQuoteIdInterface $maskedQuoteIdToQuoteId,
+        BuyRequestBuilder $requestBuilder
+    ) {
+        $this->productRepository = $productRepository;
+        $this->cartRepository = $cartRepository;
+        $this->maskedQuoteIdToQuoteId = $maskedQuoteIdToQuoteId;
+        $this->requestBuilder = $requestBuilder;
+    }
+
+    /**
+     * @inheritdoc
+     * @throws NoSuchEntityException Could not find a Cart with provided $maskedCartId
+     */
+    public function execute(string $maskedCartId, array $cartItems): AddProductsToCartOutput
+    {
+        $cartId = $this->maskedQuoteIdToQuoteId->execute($maskedCartId);
+        $cart = $this->cartRepository->get($cartId);
+
+        foreach ($cartItems as $n => $cartItem) {
+            $this->addItemToCart($cart, $cartItem, $n);
+        }
+
+        if ($cart->getData('has_error')) {
+            $errors = $cart->getErrors();
+
+            /** @var MessageInterface $error */
+            foreach ($errors as $error) {
+                $this->addError($error->getText());
+            }
+        }
+
+        $this->cartRepository->save($cart);
+
+        return $this->prepareOutput($cart);
+    }
+
+    /**
+     * @param CartInterface|Quote $cart
+     * @param Data\CartItem $cartItem
+     * @param int $cartItemPosition
+     */
+    private function addItemToCart(CartInterface $cart, Data\CartItem $cartItem, int $cartItemPosition): void
+    {
+        $sku = $cartItem->getSku();
+
+        try {
+            $product = $this->productRepository->get($sku, false, null, true);
+        } catch (NoSuchEntityException $e) {
+            $this->addError(
+                __('Could not find a product with SKU "%sku"', ['sku' => $sku])->render(),
+                $cartItemPosition,
+                self::ERROR_PRODUCT_NOT_FOUND
+            );
+
+            return;
+        }
+
+        try {
+            $result = $cart->addProduct($product, $this->requestBuilder->build($cartItem));
+        } catch (\Throwable $e) {
+            $this->addError(
+                __(
+                    'Could not add the product with SKU %sku to the shopping cart: %message',
+                    ['sku' => $sku, 'message' => $e->getMessage()]
+                )->render(),
+                $cartItemPosition
+            );
+            return;
+        }
+
+        if (is_string($result)) {
+            $errors = array_unique(explode("\n", $result));
```

Comment:
For complex errors we should be able utilize composite `GraphQlInputException`. Please see example `\Magento\QuoteGraphQl\Model\Cart\AddSimpleProductToCart::execute`.

Reply:
Hi @lenaorobei. Thank you for your suggestion. In case of GraphQl modules, this approach will work. However, in this particular case, we work with the `Quote` module that has no dependency to GraphQl modules. 
We use a similar approach in the `Reorder` functionality 

https://github.com/magento/magento2/blob/03a9a241ee14b0b9761d1e40db0370e22e999389/app/code/Magento/Sales/Model/Reorder/Reorder.php#L189